src/api.c

```diff
@@ -2455,88 +2436,119 @@ int cgroup_create_cgroup(struct cgroup *cgroup, int ignore_ownership)
-	 * data structure. If not, we fail.
-	 */
-	for (k = 0; k < cgroup->index; k++) {
-		if (!cg_build_path(cgroup->name, path,
-				cgroup->controller[k]->name))
-			continue;
-
-		error = cgroup_get_controller_version(
-			cgroup->controller[k]->name, &version);
-		if (error)
-			goto err;
-
-		if (version == CGROUP_V2) {
-			char *parent, *dname;
+	if (!cg_build_path(cgroup->name, path, controller->name))
+		goto err;
 
-			parent = strdup(path);
-			if (!parent) {
-				error = ECGOTHER;
-				goto err;
-			}
+	error = cgroup_get_controller_version(controller->name, &version);
+	if (error)
+		goto err;
 
-			dname = dirname(parent);
+	if (version == CGROUP_V2) {
+		char *parent, *dname;
 
-			error = cgroupv2_subtree_control_recursive(dname,
-					cgroup->controller[k]->name, true);
-			free(parent);
-			if (error)
-				goto err;
+		parent = strdup(path);
+		if (!parent) {
+			error = ECGOTHER;
+			goto err;
 		}
 
-		error = cg_create_control_group(path);
+		dname = dirname(parent);
+
+		error = cgroupv2_subtree_control_recursive(dname,
+				controller->name, true);
+		free(parent);
 		if (error)
 			goto err;
+	}
 
-		base = strdup(path);
+	error = cg_create_control_group(path);
+	if (error)
+		goto err;
 
-		if (!base) {
-			last_errno = errno;
-			error = ECGOTHER;
-			goto err;
-		}
+	base = strdup(path);
 
-		if (!ignore_ownership) {
-			cgroup_dbg("Changing ownership of %s\n", fts_path[0]);
-			error = cg_chown_recursive(fts_path,
-				cgroup->control_uid, cgroup->control_gid);
-			if (!error)
-				error = cg_chmod_recursive_controller(fts_path[0],
-						cgroup->control_dperm,
-						cgroup->control_dperm != NO_PERMS,
-						cgroup->control_fperm,
-						cgroup->control_fperm != NO_PERMS,
-						1, cgroup_ignored_tasks_files);
-		}
+	if (!base) {
+		last_errno = errno;
+		error = ECGOTHER;
+		goto err;
+	}
 
-		if (error)
-			goto err;
+	if (!ignore_ownership) {
+		cgroup_dbg("Changing ownership of %s\n", fts_path[0]);
+		error = cg_chown_recursive(fts_path,
+			cgroup->control_uid, cgroup->control_gid);
+		if (!error)
+			error = cg_chmod_recursive_controller(fts_path[0],
+					cgroup->control_dperm,
+					cgroup->control_dperm != NO_PERMS,
+					cgroup->control_fperm,
+					cgroup->control_fperm != NO_PERMS,
+					1, cgroup_ignored_tasks_files);
+	}
 
-		error = cgroup_set_values_recursive(base,
-				cgroup->controller[k], false);
+	if (error)
+		goto err;
+
+	error = cgroup_set_values_recursive(base, controller, false);
+	if (error)
+		goto err;
+
+	if (!ignore_ownership && version == CGROUP_V1) {
+		error = cgroup_chown_chmod_tasks(base,
+				cgroup->tasks_uid, cgroup->tasks_gid,
+				cgroup->task_fperm);
 		if (error)
 			goto err;
-
-		if (!ignore_ownership && version == CGROUP_V1) {
-			error = cgroup_chown_chmod_tasks(base,
-					cgroup->tasks_uid, cgroup->tasks_gid,
-					cgroup->task_fperm);
-			if (error)
-				goto err;
-		}
-		free(base);
-		base = NULL;
 	}
+	free(base);
+	base = NULL;
```

Comment:
@drakenclimber The patchset looks great, I gave it a try and it works as expected.  A few minor nitpicks, the above `free()` and `base = NULL` can be removed, it's free below.